lib/src/back_end/solution.dart

```diff
@@ -0,0 +1,167 @@
+    var writer = CodeWriter(pageWidth, state);
+    writer.format(root);
+    var (text, score) = writer.finish();
+    return Solution._(state, text, score);
+  }
+
+  Solution._(this._state, this.text, this.score);
+
+  /// When called on a [Solution] with some unselected piece states, chooses a
+  /// piece and yields further solutions for each state that piece can have.
+  List<Solution> expand(Piece root, int pageWidth) {
+    var piece = _state.firstUnsolved();
+    if (piece == null) return const [];
+
+    var result = <Solution>[];
+    for (var i = 0; i < piece.stateCount; i++) {
+      var solution = Solution(root, pageWidth, _state.cloneWith(piece, i));
+      result.add(solution);
+    }
+
+    return result;
+  }
+
+  /// For performance, we want to stop checking solutions as soon as we find
+  /// the best one. Best means the fewest overflow characters and the lowest
+  /// code.
+  @override
+  int compareTo(Solution other) {
+    var scoreComparison = score.compareTo(other.score);
+    if (scoreComparison != 0) return scoreComparison;
+
+    // Should be solving the same set of pieces.
+    assert(_state._pieces.length == other._state._pieces.length);
+
+    // If all else is equal, prefer lower states in earlier pieces.
+    // TODO(tall): This might not be needed once piece scoring is more
+    // sophisticated.
+    for (var i = 0; i < _state._pieces.length; i++) {
+      var piece = _state._pieces[i];
+      var thisState = _state.pieceState(piece);
+      var otherState = other._state.pieceState(piece);
+      if (thisState != otherState) return thisState.compareTo(otherState);
+    }
+
+    return 0;
+  }
+
+  @override
+  String toString() => '$score $_state';
+}
+
+class Score implements Comparable<Score> {
+  // TODO(tall): Should this actually be part of scoring? Do we want to use
+  // validity to determine how we order solutions to explore?
+  /// Whether this score is for a valid solution or not.
+  ///
+  /// An invalid solution is one where a hard newline appears in a context
+  /// where splitting isn't allowed. This is considered worse than any other
+  /// solution.
+  final bool isValid;
+
+  /// The number of characters that do not fit inside the page width.
+  final int overflow;
+
+  /// The amount of penalties applied based on the chosen line splits.
+  final int cost;
+
+  Score({required this.isValid, required this.overflow, required this.cost});
+
+  @override
+  int compareTo(Score other) {
+    // All invalid solutions are equal.
+    if (!isValid && !other.isValid) return 0;
+
+    // We are looking for *lower* costs and overflows, so an invalid score is
+    // considered higher or after all others.
+    if (!isValid) return 1;
+    if (!other.isValid) return -1;
+
+    // Overflow is always penalized more than line splitting cost.
+    if (overflow != other.overflow) return overflow.compareTo(other.overflow);
+
+    return cost.compareTo(other.cost);
```

Comment:
It's not fleshed out yet in the new IR because I don't have enough of the language working where interesting cost comes into play. With the old Chunk IR and line splitter, each Rule can have a different cost value when it's forced to split and the costs are tweaked to prefer some solutions over others. Also, you can wrap a series of chunks in a Span and whenever a split occurs between those chunks, the Span's cost is added in.

I don't know how much of that we'll need for the new IR (I'm hoping a more explicitly tree-like IR lets us do something simpler), so write now it's just "the number of Pieces whose state isn't zero". But it's not documented yet because it's not fully baked yet. I'll fill it in more when more of the language is in place.